### PR TITLE
Fix protecode comp version bug

### PR DESCRIPTION
**Pull request by jkrayl**

### CODE REVIEW

File: protecode/scanning_util.py [hunk 1]
```diff
@@ -290,15 +294,18 @@ def enumerate_reference_triages():
 
         if self._processing_mode is ProcessingMode.FORCE_UPLOAD:
             logger.info('force-upload - will re-upload all images')
-            images_to_upload += list(resource_group.resources())
+            for upload_element in upload_elements:
+                elements_to_upload.append(upload_element)
```
Comment (zkdev) [inline protecode/scanning_util.py:298]: this loop seems a bit inconvenient

File: protecode/scanning_util.py [hunk 2]
```diff
@@ -338,7 +346,8 @@ def enumerate_reference_triages():
                     'Marked existing product(s) with ID(s) '
                     f"'{','.join([str(p.product_id()) for p in protecode_apps_to_remove])}' "
                     'that had no match in the current group '
-                    f"'{resource_group.component().name}, {resource_group.image_name()}' "
+                    # TODO don't access first element
```
Comment (zkdev) [inline protecode/scanning_util.py:349]: TODO can be removed

File: protecode/scanning_util.py [hunk 3]
```diff
@@ -377,41 +387,45 @@ def enumerate_reference_triages():
...
+                logger.info(f'finished waiting for {protecode_app.product_id()}')
 
+        scan_results = list(
+            wait_for_scan_results(
+                protecode_apps_to_consider=protecode_apps_to_consider,
+            )
+        )
         # apply imported triages for all protecode apps
-        for protecode_app in protecode_apps_to_consider:
+        for protecode_app in scan_results:
```
Comment (zkdev) [inline protecode/scanning_util.py:428]: we should rename many variable and refactor some parts to simplify maintenance, but not part of this PR :)

File: protecode/util.py [hunk 4]
```diff
@@ -137,17 +138,17 @@ def _filter_resources_to_scan(component: cm.Component, resource: cm.Resource):
                 return True
 
         for _, components in component_groups.items():
```
Comment (zkdev) [inline protecode/util.py:140]: we should use `.values()` instead

**Review by zkdev — CHANGES_REQUESTED** [review]

File: protecode/scanning_util.py (hunk 1 above)
Comment (ccwienk) [reply protecode/scanning_util.py:298]: true - it does not improve readability compared to the previous version (might be slightly more efficient, though). also, since this does not change the semantics, I think this ought to be part of a refactoring-commit

File: protecode/scanning_util.py (hunk 3 above)
Comment (ccwienk) [reply protecode/scanning_util.py:428]: +1 - please keep the semantical focus of this PR on corecting the bug, and add a refactoring commit afterwards.

File: protecode/util.py (hunk 4 above)
Comment (ccwienk) [reply protecode/util.py:140]: please also do this in a separate refactoring commit

File: protecode/model.py [hunk 5]
```diff
@@ -257,7 +265,7 @@ class UploadResult:
     def __init__(
             self,
             status: UploadStatus,
-            component: Component,
+            component,
```
Comment (ccwienk) [inline protecode/model.py:268]: what's the expected type?

File: protecode/model.py [hunk 6]
```diff
@@ -32,6 +34,12 @@ class ProcessingStatus(Enum):
     FAILED = 'F'
 
 
+@dataclasses.dataclass(frozen=True)
+class UploadElement:
```
Comment (ccwienk) [inline protecode/model.py:38]: I think this does not reasonably belong to protecode's model module. Might be a better fit for either `gci.componentmodel`, or our local `cnudie` package. The name is also too strongly influenced by the use-case for which it was added. It is in fact a two-tuple of a resource and its resource. Why not simply call it ComponentResource.

for the time being, at least mv it to protecode.scanning_util

File: protecode/scanning_util.py [hunk 7]
```diff
@@ -234,30 +235,33 @@ def _metadata(
 
     def upload_container_image_group(
         self,
-        resource_group: ResourceGroup,
+        upload_elements: typing.List[UploadElement],
```
Comment (ccwienk) [inline protecode/scanning_util.py:238]: do you actually need a list? if so, you could use the shorthand form `list[T]`. if not, rather use `typing.Iterable`

File: protecode/scanning_util.py [hunk 8]
```diff
@@ -368,7 +378,7 @@ def enumerate_reference_triages():
                             f'{resource.access.imageReference=} no longer available to protecode '
                             f'- will upload. Corresponding product: {protecode_app.product_id()}'
                         )
-                        images_to_upload.append(resource)
+                        elements_to_upload.append(upload_element)
```
Comment (ccwienk) [inline protecode/scanning_util.py:381]: naming-nit: the term we use in cnudie would be either `resources`, or - covering also `sources` - `artefacts`. Please choose either `artefact` or `resource`, if doing a rename.

note that currently, the artefacts will always be `OciImages`, as we did so far not implement this for any other artefact types.

File: protecode/scanning_util.py [hunk 9]
```diff
@@ -377,41 +387,45 @@ def enumerate_reference_triages():
                 self._api.rescan(protecode_app.product_id())
 
         # upload new images
-        for resource in images_to_upload:
+        for upload_element in elements_to_upload:
             try:
                 scan_result = self._upload_image(
```
Comment (ccwienk) [inline protecode/scanning_util.py:392]: nit: if thinking about renaming the variable names (image -> _artefact_), the method names ought also be renamed. to have a chance of eventually finalising this PR, please revert those name-changes back for the time being, so we can keep the focus on correcting the bug. do refactoring later.

**Review by ccwienk — CHANGES_REQUESTED** [review]

**Comment by ccwienk** [comment]
obsolete (was split into two PRs)
